native/dlasq2.go

```diff
@@ -0,0 +1,366 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/lapack"
+)
+
+// Dlasq2 computes all the eigenvalues of the symmetric positive
+// definite tridiagonal matrix associated with the qd array Z. Eigevalues
+// are computed to high relative accuracy avoiding denormalization, underflow
+// and overflow.
+//
+// To see the relation of Z to the tridiagonal matrix, let L be a
+// unit lower bidiagonal matrix with subdiagonals Z(2,4,6,,..) and
+// let U be an upper bidiagonal matrix with 1's above and diagonal
+// Z(1,3,5,,..). The tridiagonal is L*U or, if you prefer, the
+// symmetric tridiagonal to which it is similar.
+//
+// info returns a status error. The return codes mean as follows:
+//  0: The algorithm completed successfully.
+//  1: A split was marked by a positive value in e.
+//  2: Current block of Z not diagonalized after 100*n iterations (in inner
+//     while loop). On exit Z holds a qd array with the same eigenvalues as
+//     the given Z.
+//  3: Termination criterion of outer while loop not met (program created more
+//     than N unreduced blocks).
+func (impl Implementation) Dlasq2(n int, z []float64) (info int) {
+	// TODO(btracey): make info an error.
+	if len(z) < 4*n {
+		panic(badZ)
+	}
+	const cbias = 1.5
+
+	eps := dlamchP
+	safmin := dlamchS
+	tol := eps * 100
+	tol2 := tol * tol
+	if n < 0 {
+		panic(nLT0)
+	}
+	if n == 0 {
+		return info
+	}
+	if n == 1 {
+		if z[0] < 0 {
+			panic(negZ)
+		}
+		return info
+	}
+	if n == 2 {
+		if z[1] < 0 || z[2] < 0 {
+			panic("lapack: bad z value")
+		} else if z[2] > z[0] {
+			z[0], z[2] = z[2], z[0]
+		}
+		z[4] = z[0] + z[1] + z[2]
+		if z[1] > z[2]*tol2 {
+			t := 0.5 * (z[0] - z[2] + z[1])
+			s := z[2] * (z[1] / t)
+			if s <= t {
+				s = z[2] * (z[1] / (t * (1 + math.Sqrt(1+s/t))))
+			} else {
+				s = z[2] * (z[1] / (t + math.Sqrt(t)*math.Sqrt(t+s)))
+			}
+			t = z[0] + s + z[1]
+			z[2] *= z[0] / t
+			z[0] = t
+		}
+		z[1] = z[2]
+		z[5] = z[1] + z[0]
+		return info
+	}
+	// Check for negative data and compute sums of q's and e's.
+	z[2*n-1] = 0
+	emin := z[1]
+	var d, e, qmax, zmax float64
+	var i1, n1 int
+	for k := 0; k < 2*(n-1); k += 2 {
+		if z[k] < 0 {
```

Comment:
Make it `if z[k] < 0 || z[k+1] < 0` ? Also, it is not clear (at least not to me) where the requirement that elements of z must be non-negative comes from. At least it should be mentioned in the docs.